File: octoprint_octolapse/gcode_parser.py
# coding=utf-8
##################################################################################
# Octolapse - A plugin for OctoPrint used for making stabilized timelapse videos.
# Copyright (C) 2017  Brad Hochgesang
##################################################################################
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published
# by the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see the following:
# https://github.com/FormerLurker/Octolapse/blob/master/LICENSE
#
# You can contact the author either through the git-hub repository, or at the
# following email address: FormerLurker@pm.me
##################################################################################
import operator
from six import string_types


class CommandParameter(object):
    def __init__(self, name, parse_function, order):
        self.Name = name
        self.ParseFunction = parse_function
        self.Order = order

    @staticmethod
    def parse_float_positive(parameter_string):
        value, parameters = CommandParameter.parse_float(parameter_string)
        if value is None:
            return None
        elif value < 0:
            raise ValueError("The parameter value is negative, which is not allowed.")

        return value, parameters

    @staticmethod
    def parse_float(parameter_string):
        assert (isinstance(parameter_string, string_types))
        parameter_string = parameter_string.lstrip()

        index = 0
        sign_seen = False
        period_seen = False

        float_string = ""
        for index in range(0, len(parameter_string)):
            _c = parameter_string[index]
            if _c.isspace():
                continue
            if _c in ["+", "-"]:
                if not sign_seen:
                    sign_seen = True
                    float_string += _c
                else:
                    raise ValueError("Could not parse float from parameter string, saw multiple signs.")
            elif "0" <= _c <= "9":
                float_string += _c
            elif _c == ".":
                if not period_seen:
                    float_string += _c
                    period_seen = True
                else:
                    raise ValueError("Could not parse float from parameter string, saw multiple decimal points.")
            else:
                break
        if len(parameter_string) > index:
            parameter_string = parameter_string[index:]
        else:
            parameter_string = ""

        value = None
        if len(float_string) > 0:
            value = float(float_string)

        return value, parameter_string


class Command(object):

    def __init__(self, command, name, display_template=None, parameters=None, text_only_parameter=False):
        self.Command = command
        self.Name = name
        self.DisplayTemplate = display_template
        self.Parameters = parameters
        self.TextOnlyParameter = text_only_parameter

    def parse_parameters(self, parameters_string):
        parameters = {}
        parameter = None
        parameter_cmd = None
        index = 0
        for index in range(0, len(parameters_string)):
            _c = parameters_string[index].upper()
            if _c.isspace():
                continue
            elif _c in self.Parameters:
                parameter = _c
                parameter_cmd = self.Parameters[parameter].ParseFunction

                break

        if parameter is not None and len(parameters_string) > 0:
            parameters_string = parameters_string[index + 1:]
            if parameter_cmd is not None:
                parameter_value, parameters_string = parameter_cmd(parameters_string)
                parameters[parameter] = parameter_value
                additional_parameters = self.parse_parameters(parameters_string)
                if any(filter(parameters.has_key, additional_parameters.keys())):
                    raise ValueError("Either a parameter value was repeated or an unexpected character was found, "
                                     "cannot parse gcode.")
                parameters.update(additional_parameters)

        return parameters

    def to_string(self):
        # if we do not have gcode, construct from the command and parameters
        command_string = self.Command

        # loop through each parameter and add the parameter name and value to the command string
        for parameter in (sorted(self.Parameters.values(), key=operator.attrgetter('Order'))):
            if parameter.Value is not None:
                command_string += " " + parameter.Name + str(parameter.Value)
        # since there is no gcode, we can't have a comment.  Time to return the command string
        return command_string


class Commands(object):
    G0 = Command(
        "G0",
        "Rapid linear move",
        "G0 - Linear move to X={X}, Y={Y}, Z={Z}, E={E}, F={F}",
        parameters={
            "X": CommandParameter("X", CommandParameter.parse_float, 1),
            "Y": CommandParameter("Y", CommandParameter.parse_float, 2),
            "Z": CommandParameter("Z", CommandParameter.parse_float, 3),
            "E": CommandParameter("E", CommandParameter.parse_float, 4),
            "F": CommandParameter("F", CommandParameter.parse_float_positive, 5)
        }
    )

    G1 = Command(
        "G1",
        "Linear move",
        "G1 - Linear move to X={X}, Y={Y}, Z={Z}, E={E}, F={F}",
        parameters={
            "X": CommandParameter("X", CommandParameter.parse_float, 1),
            "Y": CommandParameter("Y", CommandParameter.parse_float, 2),
            "Z": CommandParameter("Z", CommandParameter.parse_float, 3),
            "E": CommandParameter("E", CommandParameter.parse_float, 4),
            "F": CommandParameter("F", CommandParameter.parse_float_positive, 5)
        }
    )
    G2 = Command(
        "G2",
        "Clockwise Arc Move",
        "G2 - Clockwise arc move to X={X}, Y={Y}, I={I}, J={J}, R={R}, E={E}, F={F}",
        parameters={
            "X": CommandParameter("X", CommandParameter.parse_float, 1),
            "Y": CommandParameter("Y", CommandParameter.parse_float, 2),
            "I": CommandParameter("I", CommandParameter.parse_float, 3),
            "J": CommandParameter("J", CommandParameter.parse_float, 4),
            "R": CommandParameter("R", CommandParameter.parse_float, 5),
            "E": CommandParameter("E", CommandParameter.parse_float, 6),
            "F": CommandParameter("F", CommandParameter.parse_float_positive, 7)
        }
    )
    G3 = Command(
        "G3",
        "Counter-Clockwise Arc Move",
        "G3 - Counter-clockwise arc move to X={X}, Y={Y}, I={I}, J={J}, R={R}, E={E}, F={F}",
        parameters={
            "X": CommandParameter("X", CommandParameter.parse_float, 1),
            "Y": CommandParameter("Y", CommandParameter.parse_float, 2),
            "I": CommandParameter("I", CommandParameter.parse_float, 3),
            "J": CommandParameter("J", CommandParameter.parse_float, 4),
            "R": CommandParameter("R", CommandParameter.parse_float, 5),
            "E": CommandParameter("E", CommandParameter.parse_float, 6),
            "F": CommandParameter("F", CommandParameter.parse_float_positive, 7)
        }
    )
    G10 = Command(
        "G10",
        "Retract"
        "G10 - Retract or Tool Offset",
        parameters={
            "P": CommandParameter("P", CommandParameter.parse_float, 1),
            "X": CommandParameter("X", CommandParameter.parse_float, 1),
            "Y": CommandParameter("Y", CommandParameter.parse_float, 1),
            "U": CommandParameter("U", CommandParameter.parse_float, 1),
            "V": CommandParameter("V", CommandParameter.parse_float, 1),
            "W": CommandParameter("W", CommandParameter.parse_float, 1),
            "Z": CommandParameter("Z", CommandParameter.parse_float, 1),
            "R": CommandParameter("R", CommandParameter.parse_float, 1),
            "S": CommandParameter("S", CommandParameter.parse_float, 1)
        }
    )

    G11 = Command(
        "G11",
        "Retract"
        "G11 - Unretract",
        parameters={
            "S": CommandParameter("S", CommandParameter.parse_float, 1)
        }
    )
    G20 = Command(
        "G20",
        "Set units to inches"
        "G20 - Set Units to Inches",
        parameters={}
    )

    G21 = Command(
        "G21",
        "Set units to millimeters"
        "G21 - Set Units to Millimeters",
        parameters={}
    )

    G28 = Command(
        "G28",
        "Go to origin",
        "G28 - Go to origin X={X}, Y={Y}, Z={Z}, W={W}",
        parameters={
            "X": CommandParameter("X", CommandParameter.parse_float, 1),
            "Y": CommandParameter("Y", CommandParameter.parse_float, 2),
            "Z": CommandParameter("Z", CommandParameter.parse_float, 3),
            "W": CommandParameter("W", CommandParameter.parse_float, 4)
        }
    )

    G29 = Command(
        "G29",
        "Detailed Z-Prob",
        "G29 - Dtailed z-probe: S={S}",
        parameters={
            "S": CommandParameter("S", CommandParameter.parse_float, 1)
        }
    )

    G80 = Command(
        "G80",
        "Cancel canned cycle (firmware specific)"
        "G80 - Cancel canned cycle (firmware specific)",
        parameters={}
    )

    G90 = Command(
        "G90",
        "Set XYZ axes to absolute coordinates",
        "G90 - Set XYZ axes to absolute coordinates",
        parameters={}
    )
    G91 = Command(
        "G91",
        "Set XYZ axes to relative coordinates",
        "G91 - Set XYZ axes to relative coordinates",
        parameters={}
    )

    G92 = Command(
        "G92",
        "Set absolute position",
        "G02 - Set absolute position to X={X}, Y={Y}, Z={Z}, E={E}",
        parameters={
            "X": CommandParameter("X", CommandParameter.parse_float, 1),
            "Y": CommandParameter("Y", CommandParameter.parse_float, 2),
            "Z": CommandParameter("Z", CommandParameter.parse_float, 3),
            "E": CommandParameter("E", CommandParameter.parse_float, 4),
            "O": CommandParameter("O", CommandParameter.parse_float, 5)
        }
    )

    M82 = Command(
        "M82",
        "Set extruder relative mode",
        "M82 - Set extruder to relative mode",
        parameters={}
    )

    M83 = Command(
        "M83",
        "Set extruder absolute mode",
        "M83 - Set extruder to absolute mode",
        parameters={}
    )

    M104 = Command(
        "M104",
        "Set extruder temperature",
        "M104 - Set extruder temperature: S={S}",
        parameters={
            "S": CommandParameter("S", CommandParameter.parse_float, 1)
        }
    )

    M105 = Command(
        "M105",
        "Get temperature",
        "M105 - Get temperatures",
        parameters={}
    )

    M106 = Command(
        "M106",
        "Set part fan speed",
        "M106 - Set part fan speed: S={S}, P={P}, I={I}, F={F}, L={L}, B={B}, R={R}, T={T}",
        parameters={
            "S": CommandParameter("S", CommandParameter.parse_float, 1),
            "P": CommandParameter("P", CommandParameter.parse_float, 2),  # previously only supported ints
            "I": CommandParameter("I", CommandParameter.parse_float, 3),  # previously only supported booleans
            "F": CommandParameter("F", CommandParameter.parse_float, 4),  # previously only supported ints
            "L": CommandParameter("L", CommandParameter.parse_float, 5),
            "B": CommandParameter("B", CommandParameter.parse_float, 6),  # previously only supported ints
            "R": CommandParameter("R", CommandParameter.parse_float, 7),  # previously only supported ints
            "T": CommandParameter("T", CommandParameter.parse_float, 8),  # previously only supported ints
        }
    )

    M109 = Command(
        "M109",
        "Set extruder temperature and wait",
        display_template="M109 - Set extruder temperature and wait: S={S}, R={R}",
        parameters={
            "S": CommandParameter("S", CommandParameter.parse_float, 1),
            "R": CommandParameter("R", CommandParameter.parse_float, 2)
        }
    )

    M114 = Command(
        "M114",
        "Get position"
        "M114 - Get current position",
        parameters={}
    )

    M116 = Command(
        "M116",
        "Wait for temperature",
        display_template="M116 - Wait for temperature: P={P}, H={H}, C={C}",
        parameters={
            "P": CommandParameter("P", CommandParameter.parse_float, 1),
            "H": CommandParameter("H", CommandParameter.parse_float, 2),
            "C": CommandParameter("C", CommandParameter.parse_float, 3)
        }
    )

    M140 = Command(
        "M140",
        "Set bed temperature",
        "M140 - Set bed temperature: S={S}, H={H}",
        parameters={
            "S": CommandParameter("S", CommandParameter.parse_float, 1),
            "H": CommandParameter("H", CommandParameter.parse_float, 2)
        }
    )

    M141 = Command(
        "M141",
        "Set chamber temperature",
        "M141 - Set chamber temperature: S={S}, H={H}",
        parameters={
            "S": CommandParameter("S", CommandParameter.parse_float, 1),
            "H": CommandParameter("H", CommandParameter.parse_float, 2)
        }
    )

    M190 = Command(
        "M190",
        "Set bed temperature and wait",
        "M190 - Set bed temperature and wait: S={S}, R={R}",
        parameters={
            "S": CommandParameter("S", CommandParameter.parse_float, 1),
            "R": CommandParameter("R", CommandParameter.parse_float, 2)
        }
    )

    M191 = Command(
        "M191",
        "Set chamber temperature and wait",
        "M191 - Set chamber temperature and wait: S={S}, R={R}",
        parameters={
            "S": CommandParameter("S", CommandParameter.parse_float, 1),
            "R": CommandParameter("R", CommandParameter.parse_float, 2)
        }
    )

    M207 = Command(
        "M207",
        "Set retract length"
        "M207 - Set retract length",
        parameters={
            "S": CommandParameter("S", CommandParameter.parse_float, 1),
            "R": CommandParameter("R", CommandParameter.parse_float, 2),
            "F": CommandParameter("F", CommandParameter.parse_float, 3),
            "T": CommandParameter("T", CommandParameter.parse_float, 4),
            "Z": CommandParameter("Z", CommandParameter.parse_float, 5)
        }
    )

    M208 = Command(
        "M208",
        "Set detract length"
        "M208 - Set detract length",
        parameters={
            "S": CommandParameter("S", CommandParameter.parse_float, 1),
            "F": CommandParameter("F", CommandParameter.parse_float, 2),

        }
    )

    M400 = Command(
        "M400",
        "Wait for moves to finish",
        "M400 - Wait for moves to finish",
        parameters={}
    )

    CommandsDictionary = {
            G0.Command: G0,
            G1.Command: G1,
            G2.Command: G2,
            G3.Command: G3,
            G10.Command: G10,
            G11.Command: G11,
            G20.Command: G20,
            G21.Command: G21,
            G28.Command: G28,
            G29.Command: G29,
            G80.Command: G80,
            G90.Command: G90,
            G91.Command: G91,
            G92.Command: G92,
            M82.Command: M82,
            M83.Command: M83,
            M104.Command: M104,
            M105.Command: M105,
            M106.Command: M106,
            M109.Command: M109,
            M114.Command: M114,
            M116.Command: M116,
            M140.Command: M140,
            M141.Command: M141,
            M190.Command: M190,
            M191.Command: M191,
            M207.Command: M207,
            M208.Command: M208,
            M400.Command: M400
        }

    GcodeWords = {"G", "M"}
    SuppressedSavedCommands = [M105.Command, M400.Command]
    SuppressedSnapshotGcodeCommands = [M105.Command]
    CommandsRequireMetric = [G0.Command, G1.Command, G2.Command, G3.Command, G28.Command, G92.Command]
    TestModeSuppressExtrusionCommands = [G0.Command, G1.Command, G2.Command, G3.Command]
    TestModeSuppressCommands = [
        G10.Command, G11.Command,
        M104.Command, M140.Command, M141.Command,
        M109.Command, M190.Command, M191.Command,
        M116.Command, M106.Command
    ]

    @staticmethod
    def parse(gcode):
        # strip off any trailing comments
        ix = gcode.find(";")
        if ix > -1:
            if ix <= len(gcode) - 1:
                gcode = gcode[0:ix]
            else:
                return None, None

        # remove any comments from ('s
        start_comment_index = None
        end_comment_index = None

        index = 0
        while index < len(gcode):
            if end_comment_index is not None:
                # we have a comment to remove.
                if start_comment_index == 0:
                    if len(gcode) == end_comment_index + 1:
                        gcode = ""
                        index = 0
                    else:
                        gcode = gcode[end_comment_index + 1:]
                        index = 0

                elif end_comment_index + 1 == len(gcode):
                    gcode = gcode[0:start_comment_index]
                    index = start_comment_index + 1
                else:
                    gcode = gcode[0:start_comment_index] + gcode[end_comment_index + 1:]
                    index = start_comment_index

                start_comment_index = None
                end_comment_index = None

            for index in range(index, len(gcode)):
                c = gcode[index]
                if start_comment_index is None and c == "(":
                    start_comment_index = index
                elif start_comment_index is not None and c == ")":
                    end_comment_index = index
                    break

            if start_comment_index is None or end_comment_index is None:
                break

        # strip whitespace
        gcode = gcode.strip()

        # ignore blank lines
        if len(gcode) < 1:
            return None, None

        # ignore any lines that start with a %
        if gcode[0] == "%":
            return None, None

        # make sure our string is greater than 2 characters
        if len(gcode) < 2:
            return None, None

        # extract any line numbers
        if gcode[0] == "N":
            _n_temp = ""
            # extract any integers and ignore whitespace
            for index in range(0, len(gcode) - 1):
                _c = gcode[index].upper()

                if _c.isspace():
                    continue
                if "0" <= _c <= "9":
                    _n_temp += _c
                else:
                    break

            # remove the line number from the command string and strip off whitespace from the front
            gcode = gcode[index:].lstrip()

        # Now we should be left with the command and any parameters
        # Make sure our command is a valid one
        if len(gcode) < 2:
            return None, None

        # get the command letter
        command_letter = gcode[0].upper()
        if command_letter not in Commands.GcodeWords:
            return None, None

        # search for decimals or periods to build the command address
        command_address = ""
        has_seen_period = False
        for index in range(1, len(gcode)):
            c = gcode[index]
            if c.isspace():
                continue
            elif "0" <= c <= "9":
                command_address += c
            elif c == ".":
                if not has_seen_period:
                    # if the command add
                    command_address = str(int(command_address))
                    command_address += c
                    has_seen_period = True
                else:
                    raise ValueError("Cannot parse the gcode address, multiple periods seen.")
            else:
                break
        # If we've not seen any periods, strip any leading 0s from the gcode
        if not has_seen_period:
            command_address = str(int(command_address))

        # make sure the command is in the dictionary
        command_to_search = command_letter + command_address
        if command_to_search not in Commands.CommandsDictionary.keys():
            return command_to_search, None

        cmd = Commands.CommandsDictionary[command_to_search]

        # get the parameter string
        if len(gcode) > index:
            parameters = gcode[index:]
        else:
            parameters = ""

        if not cmd.TextOnlyParameter:
            parameters = cmd.parse_parameters(parameters)

        return command_to_search, parameters

    @staticmethod
    def to_string(cmd, parameters):
        if cmd is None:
            return ""
        gcode = cmd

        if parameters is not None:
            for key, value in parameters.items():
                gcode += " " + key + str(value)
        return gcode

    @staticmethod
    def alter_for_test_mode(command_string, cmd, parameters, return_string=False):
        if cmd is None:
            if return_string:
                return command_string
            return None

        if cmd in Commands.TestModeSuppressExtrusionCommands and "E" in parameters:
            parameters.pop("E")
            # reform the gcode
            gcode = Commands.to_string(cmd, parameters)
            if return_string:
                return gcode
            return gcode,
        elif cmd in Commands.TestModeSuppressCommands:
            if return_string:
                return ""
            return None,
        else:
            if return_string:
                # reform the gcode
                return command_string
            return None

File: octoprint_octolapse/data/lib/c/GcodeParser.cpp
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// Octolapse - A plugin for OctoPrint used for making stabilized timelapse videos.
// Copyright(C) 2019  Brad Hochgesang
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// This program is free software : you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published
// by the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.If not, see the following :
// https ://github.com/FormerLurker/Octolapse/blob/master/LICENSE
//
// You can contact the author either through the git - hub repository, or at the
// following email address : FormerLurker@pm.me
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
#include "GcodeParser.h"
#include <cmath>
#include <iostream>
gcode_parser::gcode_parser()
{
	// doesn't work in the ancient version of c++ I am forced to use :(
	// or at least I don't know how to us a newer one with python 2.7
	// help...
	/* 
	std::vector<std::string> text_only_function_names = { "M117" }; // "M117" is an example of a command that would work here.

	std::vector<std::string> parsable_command_names = {
		"G0","G1","G2","G3","G10","G11","G20","G21","G28","G29","G80","G90","G91","G92","M82","M83","M104","M105","M106","M109","M114","M116","M140","M141","M190","M191","M207","M208","M240","M400","T"
	};
	*/
	// Have to resort to barbarity.
	// Text only function names
	std::vector<std::string> text_only_function_names;
	text_only_function_names.push_back(("M117"));
	// parsable_command_names
	std::vector<std::string> parsable_command_names;
	parsable_command_names.push_back("G0");
	parsable_command_names.push_back("G1");
	parsable_command_names.push_back("G2");
	parsable_command_names.push_back("G3");
	parsable_command_names.push_back("G10");
	parsable_command_names.push_back("G11");
	parsable_command_names.push_back("G20");
	parsable_command_names.push_back("G21");
	parsable_command_names.push_back("G28");
	parsable_command_names.push_back("G29");
	parsable_command_names.push_back("G80");
	parsable_command_names.push_back("G90");
	parsable_command_names.push_back("G91");
	parsable_command_names.push_back("G92");
	parsable_command_names.push_back("M82");
	parsable_command_names.push_back("M83");
	parsable_command_names.push_back("M104");
	parsable_command_names.push_back("M105");
	parsable_command_names.push_back("M106");
	parsable_command_names.push_back("M109");
	parsable_command_names.push_back("M114");
	parsable_command_names.push_back("M116");
	parsable_command_names.push_back("M140");
	parsable_command_names.push_back("M141");
	parsable_command_names.push_back("M190");
	parsable_command_names.push_back("M191");
	parsable_command_names.push_back("M207");
	parsable_command_names.push_back("M208");
	parsable_command_names.push_back("M240");
	parsable_command_names.push_back("M400");
	parsable_command_names.push_back("T");

	for (unsigned int index = 0; index < text_only_function_names.size(); index++)
	{
		std::string functionName = text_only_function_names[index];
		text_only_functions_.insert(functionName);
	}

	for (unsigned int index = 0; index < parsable_command_names.size(); index++)
	{
		std::string commandName = parsable_command_names[index];
		parsable_commands_.insert(commandName);
	}

}

gcode_parser::gcode_parser(const gcode_parser &source)
{
	// Private copy constructor - you can't copy this class
}

gcode_parser::~gcode_parser()
{
	text_only_functions_.clear();
	parsable_commands_.clear();
}

bool gcode_parser::is_gcode_word(char c)
{
	for (unsigned int index = 0; index < GCODE_WORDS.size(); ++index)
	{
		if (GCODE_WORDS[index] == c)
			return true;
	}
	return false;
}

std::string gcode_parser::strip_gcode(std::string gcode)
{
	std::string output;
	const int buffer_size = 64;
	output.reserve(buffer_size);
	const int gcode_length = gcode.size();
	for (int index = 0; index < gcode_length; ++index)
	{
		const char currentChar = gcode[index];
		if (currentChar == ';')
			return output;
		else if (isspace(currentChar) || currentChar == '\r' || currentChar == '\n')
			continue;
		else {
			if (currentChar >= 97 && currentChar <= 122)
			{
				output.append(1, gcode[index] - 32);
			}
			else
				output.append(1, currentChar);
		}

	}
	return output;
}

std::string gcode_parser::strip_new_lines(std::string gcode)
{
	int index = gcode.length();
	for (; index >-1; index--)
	{
		if (!(gcode[index] == '\n' || gcode[index] == '\r'))
			break;
	}
	if (index > -1)
	{
		return gcode.substr(0, index + 1);
	}
	else
	{
		return gcode;
	}

}

int gcode_parser::get_double_end_index(std::string gcode, int start_index)
{
	bool hasSeenPeriod = false;
	bool hasSeenPlusOrMinus = false;
	char curLetter;
	unsigned int stringLength = gcode.size();
	for (; static_cast<unsigned int>(start_index) < stringLength; ++start_index)
	{
		curLetter = gcode[start_index];
		if ('0' <= curLetter && curLetter <= '9')
			continue;
		else if (curLetter == '+' || curLetter == '-')
		{

			if (!hasSeenPlusOrMinus)
			{
				hasSeenPlusOrMinus = true;
				continue;
			}
			else
			{
				start_index = -1;
				break;
			}
		}
		else if (gcode[start_index] == '.')
		{
			if (!hasSeenPeriod)
			{
				hasSeenPeriod = true;
				continue;
			}
			else
			{
				start_index = -1;
				break;
			}

		}
		else
		{
			break;
		}

	}

	return start_index;
}

void gcode_parser::parse_gcode(std::string gcode, parsed_command * command)
{
	command->gcode = gcode;
	std::string strippedCommand = strip_gcode(gcode);
	if (strippedCommand.size() == 0 || !is_gcode_word(strippedCommand[0]))
	{
		if (strippedCommand.size() == 0)
			return;
	}
	command->cmd.append(strippedCommand, 0, 1);
	int endAddressIndex;

	if (command->cmd == "T")
	{
		endAddressIndex = 0;
	}
	else
	{
		endAddressIndex = get_double_end_index(strippedCommand, 1);
		if (endAddressIndex > 1)
		{
			command->cmd.append(strippedCommand, 1, endAddressIndex - 1);
		}
	}
	if (parsable_commands_.find(command->cmd) != parsable_commands_.end() && strippedCommand.size() > endAddressIndex)
	{
		if (text_only_functions_.find(command->cmd) != text_only_functions_.end())
		{
			get_text_only_parameter(command->cmd, gcode, &command->parameters);
		}
		else
		{
			get_parameters(strippedCommand, endAddressIndex, &command->parameters);
		}
	}
}

void gcode_parser::get_text_only_parameter(std::string command_name, std::string gcode_param, std::vector<parsed_command_parameter*>* parameters)
{
	parsed_command_parameter * param = new parsed_command_parameter();
	param->name = "TEXT";
	param->value_type = 'T';
	unsigned int textIndex;
	unsigned int foundCount = 0;
	int textStartIndex = -1;
	bool skippedSpace = false;
	for (textIndex = 0; textIndex < gcode_param.size(); textIndex++)
	{
		if (textStartIndex < 0)
		{
			if (gcode_param[textIndex] == command_name[foundCount])
			{
				foundCount++;
				if (foundCount == command_name.size())
				{
					textStartIndex = textIndex + 1;
					continue;
				}
			}
		}
		else if (!skippedSpace)
		{
			if (gcode_param[textIndex] == ' ')
				textStartIndex++;
			skippedSpace = true;
		}
		else if (gcode_param[textIndex] == ';' || gcode_param[textIndex] == '\r' || gcode_param[textIndex] == '\n')
		{
			break;
		}

	}
	std::string text = "";
	if (textIndex - textStartIndex > 0)
	{
		param->string_value = gcode_param.substr(textStartIndex, textIndex - textStartIndex);
	}

	parameters->push_back(param);

}

double gcode_parser::parse_double(const char * p) {
	double r = 0.0;
	bool neg = false;
	if (*p == '-') {
		neg = true;
		++p;
	}
	else if (*p == '+') {
		++p;
	}
	while (*p >= '0' && *p <= '9') {
		r = (r*10.0) + (*p - '0');
		++p;
	}
	if (*p == '.') {
		double f = 0.0;
		int n = 0;
		++p;
		while (*p >= '0' && *p <= '9') {
			f = (f*10.0) + (*p - '0');
			++p;
			++n;
		}
		r += f / std::pow(10.0, n);
	}
	if (neg) {
		r = -r;
	}
	return r;
}
double gcode_parser::parse_double(std::string value) {
	double r = 0.0;
	bool neg = false;
	unsigned int index = 0;
	if (value[index] == '-') {
		neg = true;
		index++;
	}
	else if (value[index] == '+') {
		index++;
	}
	while (value[index] >= '0' && value[index] <= '9') {
		r = (r*10.0) + (value[index] - '0');
		index++;
	}
	if (value[index] == '.') {
		double f = 0.0;
		int n = 0;
		index++;
		while (value[index] >= '0' && value[index] <= '9') {
			f = (f*10.0) + (value[index] - '0');
			index++;
			++n;
		}
		r += f / std::pow(10.0, n);
	}
	if (neg) {
		r = -r;
	}
	return r;
}

void gcode_parser::get_parameters(std::string commandString, int startIndex, std::vector<parsed_command_parameter*> * parameters)
{

	while ((unsigned int)startIndex < commandString.size())
	{
		parsed_command_parameter* p_param = new parsed_command_parameter();
		p_param->name.append(commandString, startIndex, 1);

		int endParameterIndex = 2;

		bool isTCommand = startIndex == 0 && commandString[0] == 'T';
		if (!isTCommand)
		{
			endParameterIndex = get_double_end_index(commandString, startIndex + 1);
		}

		if (endParameterIndex < 0 || endParameterIndex - (startIndex + 1) < 1)
		{
			p_param->value_type = 'N';
		}
		else
		{
			std::string value;
			value.append(commandString, startIndex + 1, endParameterIndex - startIndex - 1);
			if (!isTCommand)
			{
				// See if it's a double or not
				bool is_valid;
				double paramValue = 0.0;
				try
				{
					//parsedDouble = std::atof(value.c_str());
					paramValue = parse_double(value.c_str());
					is_valid = true;
				}
				catch (std::exception)
				{
					is_valid = false;
				}
				if (is_valid)
				{
					p_param->double_value = paramValue;
					p_param->value_type = 'F';
				}
				else
				{
					p_param->string_value = value;
					p_param->value_type = 'S';
				}

			}
			else
			{
				p_param->string_value = value;
				p_param->value_type = 'S';
			}
		}
		parameters->push_back(p_param);
		// Find more parameters
		startIndex = endParameterIndex;

	}
}
